beacon-chain/db/kv/regen_historical_states.go

```diff
@@ -0,0 +1,180 @@
+		if err != nil {
+			return err
+		}
+
+		// Replay blocks and replay slots if necessary.
+		if len(blocks) > 0 {
+			for i := 0; i < len(blocks); i++ {
+				if currentState.Slot() >= targetSlot {
+					break
+				}
+				if blocks[i].Block.Slot == 0 {
+					continue
+				}
+				currentState, err = regenHistoricalStateTransition(ctx, currentState, blocks[i])
+				if err != nil {
+					return err
+				}
+			}
+		}
+		if targetSlot > currentState.Slot() {
+			currentState, err = regenHistoricalStateProcessSlots(ctx, currentState, targetSlot)
+			if err != nil {
+				return err
+			}
+		}
+
+		if len(blocks) > 0 {
+			// Save the historical root, state and highest index to the DB.
+			if helpers.IsEpochStart(currentState.Slot()) && currentState.Slot()%slotsPerArchivedPoint == 0 && blocks[len(blocks)-1].Block.Slot&slotsPerArchivedPoint == 0 {
+				lastBlocksRoot, err := ssz.HashTreeRoot(blocks[len(blocks)-1].Block)
+				if err != nil {
+					return nil
+				}
+				if err := kv.SaveState(ctx, currentState, lastBlocksRoot); err != nil {
+					return err
+				}
+				if err := kv.SaveArchivedPointRoot(ctx, lastBlocksRoot, i); err != nil {
+					return err
+				}
+				if err := kv.SaveLastArchivedIndex(ctx, i); err != nil {
+					return err
+				}
+				log.WithFields(log.Fields{
+					"currentArchivedIndex/totalArchivedIndices": fmt.Sprintf("%d/%d", i, lastSavedBlockArchivedIndex),
+					"archivedStateSlot":                         currentState.Slot(),
+				}).Info("Saved historical state")
+			}
+		}
+		startSlot += slotsPerArchivedPoint
+	}
+
+	return nil
+}
+
+// This runs state transition to recompute historical state.
+func regenHistoricalStateTransition(
+	ctx context.Context,
+	state *stateTrie.BeaconState,
+	signed *ethpb.SignedBeaconBlock,
+) (*stateTrie.BeaconState, error) {
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+	if signed == nil || signed.Block == nil {
+		return nil, errors.New("block can't be nil")
+	}
+	ctx, span := trace.StartSpan(ctx, "db.regenHistoricalStateTransition")
+	defer span.End()
+	var err error
+	state, err = regenHistoricalStateProcessSlots(ctx, state, signed.Block.Slot)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not process slot")
+	}
+	state, err = transition.ProcessBlockForStateRoot(ctx, state, signed)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not process block")
+	}
+	return state, nil
+}
+
+// This runs slot transition to recompute historical state.
+func regenHistoricalStateProcessSlots(ctx context.Context, state *stateTrie.BeaconState, slot uint64) (*stateTrie.BeaconState, error) {
+	ctx, span := trace.StartSpan(ctx, "db.regenHistoricalStateProcessSlots")
+	defer span.End()
+	if state == nil {
+		return nil, errors.New("state can't be nil")
+	}
+	if state.Slot() > slot {
+		err := fmt.Errorf("expected state.slot %d < slot %d", state.Slot(), slot)
+		return nil, err
+	}
+	if state.Slot() == slot {
+		return state, nil
+	}
+	for state.Slot() < slot {
```

Comment:
This logic is pretty similar to the state transition function. To reduce code duplication, is there a way we can abstract this core logic ?